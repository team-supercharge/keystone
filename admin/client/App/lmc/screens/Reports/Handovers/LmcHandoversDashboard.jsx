import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { ActionCreators } from '../../../actions/actions'
import { connect } from 'react-redux'
import { BlankState } from '../../../../elemental'
import Swal from 'sweetalert2'
import LmcSpinner from '../../../components/LmcSpinner.jsx'
import LmcHandoversHistory from './LmcHandoversHistory.jsx'
import LmcCurrentHandover from './LmcCurrentHandover.jsx';

export class LmcHandoversDashboard extends Component {
    componentDidMount () {
        const { fetchCurrentHandoverLogs, fetchCurrentHandoverNotes } = this.props
        fetchCurrentHandoverLogs()
        fetchCurrentHandoverNotes()
    }

    showProAlert = () => {
        Swal.fire({
            title: 'Whoops!',
            html: 'You need to enable PRO mode to access this feature.<br>For more information, <a href="https://logmycare.co.uk/pricing/" target="_blank">click here.</a>',
            type: 'info',
            confirmButtonClass: 'lmc-custom-notification-confirm',
            onClose: () => this.props.router.goBack()
        })
    }

    render () {
        const { 
            handoverHistory, 
            currentHandoverLogs, 
            currentHandoverNotes 
        } = this.props

        if (!Keystone.user.features.handovers) {
            return <div>{ this.showProAlert() }</div>
        }
        if (!handoverHistory || !currentHandoverLogs || !currentHandoverNotes) {
            return <div><LmcSpinner /></div>
        }
        if (!handoverHistory.length && 
            !currentHandoverLogs.length && 
            !currentHandoverNotes.length) {
                return (
                    <div>
                        <BlankState
                            heading='No handovers information found...'
                        />
                    </div>
                )
        }
        return (
            <div>
                <LmcCurrentHandover
                    logs={currentHandoverLogs}
                    notes={currentHandoverNotes}
                />
                <LmcHandoversHistory
                    handovers={handoverHistory}
                />
            </div>
        )
    }
}

LmcHandoversDashboard.propTypes = {
    currentHandoverLogs: PropTypes.array,
    currentHandoverNotes: PropTypes.array,
    fetchCurrentHandoverLogs: PropTypes.func.isRequired,
    fetchCurrentHandoverNotes: PropTypes.func.isRequired,
    handoverHistory: PropTypes.array
}

const mapStateToProps = (state) => {
    return {
        currentHandoverLogs: state.handovers.currentLogs,
        currentHandoverNotes: state.handovers.currentNotes,
        handoverHistory: state.data.handovers,
    }
}

const mapDispatchToProps = (dispatch) => {
    return {
        fetchCurrentHandoverLogs: () => dispatch(ActionCreators.fetchCurrentHandoverLogs()),
        fetchCurrentHandoverNotes: () => dispatch(ActionCreators.fetchCurrentHandoverNotes())
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(LmcHandoversDashboard)