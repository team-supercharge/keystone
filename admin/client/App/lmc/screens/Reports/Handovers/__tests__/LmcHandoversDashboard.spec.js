import React from 'react'
import { shallow } from 'enzyme'
import { LmcHandoversDashboard } from '../LmcHandoversDashboard.jsx'

describe('LmcHandoversDashboard', () => {
    let wrapper
    let currentHandover
    let handoverHistory
    const fetchCurrentHandover = jest.fn()
    global.Keystone = { 
        adminPath: '/admin', 
        user: { features: { handovers: true } } 
    }

    beforeEach(() => {
        handoverHistory = [
            { id: 'TestId', createdBy: 'TestCarer', witnessedBy: 'TestCarer2' }
        ]
        currentHandover = {
            logs: [{ id: 'TestId', note: 'This is a note', createdBy: 'TestCarer' }],
            notes: [{ id: 'TestId', description: 'This is a log', createdBy: 'TestCarer' }]
        }

        wrapper = shallow(
            <LmcHandoversDashboard 
                handoverHistory={handoverHistory}
                currentHandover={currentHandover}
                fetchCurrentHandover={fetchCurrentHandover}
            />
        )
    })

    it('renders correctly', () => {
        expect(wrapper).toMatchSnapshot()
    })

    it('renders a BlankState if there is no data to show', () => {
        const emptyWrapper = shallow(
            <LmcHandoversDashboard 
                handoverHistory={[]}
                currentHandover={{ logs: [], notes: [] }}
                fetchCurrentHandover={fetchCurrentHandover}
            />)
        
        expect(wrapper.find('BlankState').length).toEqual(0)
        expect(emptyWrapper.find('BlankState').length).toEqual(1)
    })

    it('renders the current handover and handover history', () => {
        expect(wrapper.find('LmcCurrentHandover').length).toEqual(1)
        expect(wrapper.find('LmcHandoversHistory').length).toEqual(1)
    })
})