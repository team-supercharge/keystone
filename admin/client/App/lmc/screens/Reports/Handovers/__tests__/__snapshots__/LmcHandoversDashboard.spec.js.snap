// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LmcHandoversDashboard renders correctly 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div
    style={
      Object {
        "height": "85vh",
        "overflow": "auto",
        "padding": "50px 20px 100px 20px",
      }
    }
  >
    <div
      style={
        Object {
          "margin": "0 auto",
          "maxWidth": 1000,
          "wordWrap": "break-word",
        }
      }
    >
      <LmcCurrentHandover
        notes={
          Array [
            Object {
              "createdBy": "TestCarer",
              "description": "This is a log",
              "id": "TestId",
            },
          ]
        }
      />
      <LmcHandoversHistory
        handovers={
          Array [
            Object {
              "createdBy": "TestCarer",
              "id": "TestId",
              "witnessedBy": "TestCarer2",
            },
          ]
        }
      />
    </div>
  </div>,
  "nodes": Array [
    <div
      style={
        Object {
          "height": "85vh",
          "overflow": "auto",
          "padding": "50px 20px 100px 20px",
        }
      }
    >
      <div
        style={
          Object {
            "margin": "0 auto",
            "maxWidth": 1000,
            "wordWrap": "break-word",
          }
        }
      >
        <LmcCurrentHandover
          notes={
            Array [
              Object {
                "createdBy": "TestCarer",
                "description": "This is a log",
                "id": "TestId",
              },
            ]
          }
        />
        <LmcHandoversHistory
          handovers={
            Array [
              Object {
                "createdBy": "TestCarer",
                "id": "TestId",
                "witnessedBy": "TestCarer2",
              },
            ]
          }
        />
      </div>
    </div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <LmcHandoversDashboard
        currentHandover={
          Object {
            "logs": Array [
              Object {
                "createdBy": "TestCarer",
                "id": "TestId",
                "note": "This is a note",
              },
            ],
            "notes": Array [
              Object {
                "createdBy": "TestCarer",
                "description": "This is a log",
                "id": "TestId",
              },
            ],
          }
        }
        fetchCurrentHandover={[MockFunction]}
        handoverHistory={
          Array [
            Object {
              "createdBy": "TestCarer",
              "id": "TestId",
              "witnessedBy": "TestCarer2",
            },
          ]
        }
      />,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": LmcHandoversDashboard {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "currentHandover": Object {
            "logs": Array [
              Object {
                "createdBy": "TestCarer",
                "id": "TestId",
                "note": "This is a note",
              },
            ],
            "notes": Array [
              Object {
                "createdBy": "TestCarer",
                "description": "This is a log",
                "id": "TestId",
              },
            ],
          },
          "fetchCurrentHandover": [MockFunction],
          "handoverHistory": Array [
            Object {
              "createdBy": "TestCarer",
              "id": "TestId",
              "witnessedBy": "TestCarer2",
            },
          ],
        },
        "refs": Object {},
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <div
          style={
            Object {
              "height": "85vh",
              "overflow": "auto",
              "padding": "50px 20px 100px 20px",
            }
          }
        >
          <div
            style={
              Object {
                "margin": "0 auto",
                "maxWidth": 1000,
                "wordWrap": "break-word",
              }
            }
          >
            <LmcCurrentHandover
              notes={
                Array [
                  Object {
                    "createdBy": "TestCarer",
                    "description": "This is a log",
                    "id": "TestId",
                  },
                ]
              }
            />
            <LmcHandoversHistory
              handovers={
                Array [
                  Object {
                    "createdBy": "TestCarer",
                    "id": "TestId",
                    "witnessedBy": "TestCarer2",
                  },
                ]
              }
            />
          </div>
        </div>,
        "_debugID": 2,
        "_renderedOutput": <div
          style={
            Object {
              "height": "85vh",
              "overflow": "auto",
              "padding": "50px 20px 100px 20px",
            }
          }
        >
          <div
            style={
              Object {
                "margin": "0 auto",
                "maxWidth": 1000,
                "wordWrap": "break-word",
              }
            }
          >
            <LmcCurrentHandover
              notes={
                Array [
                  Object {
                    "createdBy": "TestCarer",
                    "description": "This is a log",
                    "id": "TestId",
                  },
                ]
              }
            />
            <LmcHandoversHistory
              handovers={
                Array [
                  Object {
                    "createdBy": "TestCarer",
                    "id": "TestId",
                    "witnessedBy": "TestCarer2",
                  },
                ]
              }
            />
          </div>
        </div>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <LmcHandoversDashboard
    currentHandover={
      Object {
        "logs": Array [
          Object {
            "createdBy": "TestCarer",
            "id": "TestId",
            "note": "This is a note",
          },
        ],
        "notes": Array [
          Object {
            "createdBy": "TestCarer",
            "description": "This is a log",
            "id": "TestId",
          },
        ],
      }
    }
    fetchCurrentHandover={[MockFunction]}
    handoverHistory={
      Array [
        Object {
          "createdBy": "TestCarer",
          "id": "TestId",
          "witnessedBy": "TestCarer2",
        },
      ]
    }
  />,
}
`;
